### PR TITLE
Make MLflow run thread-safe

### PR DESCRIPTION
**Pull request by WeichenXu123**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/WeichenXu123/mlflow/pull/13456?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/13456/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 13456
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

Make MLflow run thread-safe

Design: https://docs.google.com/document/d/1UrIj_nTyZlqQRj3_q-NYLwMeCdHnZGR-AIi-eii0oRo/edit?tab=t.0

Changes in this PR:

* Make `set_experiment` thread safe and cross processes safe and fix race conditions.
* Make MLflow active run thread local, and prevent forked subprocess inheriting active run in parent process.
* Make subprocess inherits active experiment / tracking URI / registry URI from parent process.

### How is this PR tested?

- [X] Existing unit/integration tests
- [ ] New unit/integration tests
- [X] Manual tests

I built a package for easier testing:

https://ml-team-public-read.s3.us-west-2.amazonaws.com/weichen-temp/mlflow-2.17.2.dev0-py3-none-any.whl

#### Manual test 1: Optuna mutiple-thread mode

```
import optuna
import mlflow

mlflow.set_tracking_uri("sqlite:////tmp/tune.db")
mlflow.set_experiment("tune1")


with mlflow.start_run() as parent_run:
    def objective(trial):
        with mlflow.start_run(parent_run_id=parent_run.info.run_id):
            x = trial.suggest_float('x', -10, 10)
            mlflow.log_param("x", x)
            loss = (x - 2) ** 2
            mlflow.log_metric("loss", loss)
            return loss

    study = optuna.create_study()
    # NOTE: we need to specify `n_jobs` argument otherwise
    #  no new threads are created and tuning tasks are executed
    #  within the same main thread.
    study.optimize(objective, n_trials=8, n_jobs=4)

    print("best param:", study.best_params)
```

#### Manual test 2: Optuna mutiple-process mode
Ref doc: https://optuna.readthedocs.io/en/v2.0.0/tutorial/distributed.html

```
import subprocess

import optuna
import mlflow
import multiprocessing
import uuid

mlflow.set_tracking_uri("sqlite:////tmp/tune.db")
mlflow.set_experiment("tune2")


# optuna trial storage db path
trial_db = "sqlite:////tmp/tune1.db"
study_name = f"distributed-{uuid.uuid4().hex}"

subprocess.check_call(
    f"optuna create-study --study-name {study_name} --storage {trial_db}",
    shell=True
)

with mlflow.start_run() as parent_run:
    def objective(trial):
        with mlflow.start_run(parent_run_id=parent_run.info.run_id):
            x = trial.suggest_float('x', -10, 10)
            mlflow.log_param("x", x)
            loss = (x - 2) ** 2
            mlflow.log_metric("loss", loss)
            return loss

    def subproc_fn():
        study = optuna.load_study(
            study_name=study_name, storage=trial_db
        )
        study.optimize(objective, n_trials=8)

    mp_ctx = multiprocessing.get_context('fork')

    subprocs = [
        mp_ctx.Process(target=subproc_fn)
        for _ in range(4)
    ]

    for subproc in subprocs:
        subproc.start()

    for subproc in subprocs:
        subproc.join()

    for subproc in subprocs:
        assert subproc.exitcode == 0
```

Above examples works well and logged runs are organised as nested runs, we can view them from UI like:
image

### Does this PR require documentation update?

- [X] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.

Make MLflow run thread-safe

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [X] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [X] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/tracking/fluent.py [hunk 1]
```diff
@@ -141,38 +162,50 @@ def set_experiment(
...
-            experiment_id = client.create_experiment(experiment_name)
+
+    with _experiment_lock:
+        if experiment_id is None:
+            experiment = client.get_experiment_by_name(experiment_name)
+            if not experiment:
+                try:
+                    experiment_id = client.create_experiment(experiment_name)
+                    _logger.info(
+                        "Experiment with name '%s' does not exist. Created a new experiment.",
```
Comment (BenWilson2) [inline mlflow/tracking/fluent.py:173]:
```suggestion
                        "Experiment with name '%s' does not exist. Creating a new experiment.",
```

**Review by BenWilson2 — APPROVED** [review]
The approach looks solid. Let's make sure that we file a follow-up ticket (next sprint) to add some examples and explanation to the docs about this new behavior for threading / multiprocess behavior (providing simple examples that teach the concept that are similar to what you've done for the unit tests here is ideal)

File: mlflow/tracking/_model_registry/utils.py [hunk 2]
```diff
@@ -76,6 +76,8 @@ def set_registry_uri(uri: str) -> None:
     """
     global _registry_uri
     _registry_uri = uri
+    if uri:
+        MLFLOW_REGISTRY_URI.set(_registry_uri)
```
Comment (harupy) [inline mlflow/tracking/_model_registry/utils.py:80]: Why do we need to change the environment variable?

File: mlflow/utils/thread_utils.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
+import os
+import threading
+
+_thread_locals = threading.local()
+_globals = {}
+_globals_lock = threading.Lock()
```
Comment (harupy) [inline mlflow/utils/thread_utils.py:6]: What are these? Where are these used?

File: mlflow/utils/thread_utils.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
+import os
+import threading
+
+_thread_locals = threading.local()
```
Comment (harupy) [inline mlflow/utils/thread_utils.py:4]: `locals` sounds like it's a list. Can we use `local`?

File: mlflow/utils/thread_utils.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+_globals = {}
+_globals_lock = threading.Lock()
+
+
+def set_thread_local_var(key, value):
```
Comment (harupy) [inline mlflow/utils/thread_utils.py:9]: Why do we need to set this dynamically?

File: tests/tracking/fluent/test_fluent.py [hunk 6]
```diff
@@ -1484,3 +1489,171 @@ def test_registry_uri_from_spark_conf(spark_session_with_registry_uri):
...
+
+
+def test_set_experiment_thread_safety(tmp_path):
+    sqlite_uri = "sqlite:///{}".format(tmp_path.joinpath("test.db"))
+    mlflow.set_tracking_uri(sqlite_uri)
+
+    origin_create_experiment = MlflowClient.create_experiment
+
+    def patched_create_experiment(self, *args, **kwargs):
+        time.sleep(2)
```
Comment (harupy) [inline tests/tracking/fluent/test_fluent.py:1501]: can we shorten the sleep length? 2 seconds is long.

File: mlflow/utils/thread_utils.py (hunk 3 above)
Comment (WeichenXu123) [reply mlflow/utils/thread_utils.py:6]: oh sorry. These are not used. I will clean them .

File: mlflow/utils/thread_utils.py (hunk 5 above)
Comment (WeichenXu123) [reply mlflow/utils/thread_utils.py:9]: this is a utility function to set thread local variable value. we need to change  thread local variable sometimes

**Comment by harupy** [comment]
@WeichenXu123 can you run cross-version tests and nothing will break?

**Pull request merged**